settings/generic_setting.go

```diff
@@ -0,0 +1,280 @@
+	deleteFunc func(ctx context.Context, w *databricks.AccountClient, etag string) (string, error)
+}
+
+func (w accountSetting[T]) SettingStruct() T {
+	return w.settingStruct
+}
+func (w accountSetting[T]) Read(ctx context.Context, c *databricks.AccountClient, etag string) (*T, error) {
+	return w.readFunc(ctx, c, etag)
+}
+func (w accountSetting[T]) Update(ctx context.Context, c *databricks.AccountClient, t T) (string, error) {
+	return w.updateFunc(ctx, c, t)
+}
+func (w accountSetting[T]) Delete(ctx context.Context, c *databricks.AccountClient, etag string) (string, error) {
+	return w.deleteFunc(ctx, c, etag)
+}
+func (w accountSetting[T]) GetETag(t *T) string {
+	return getEtag(t)
+}
+func (w accountSetting[T]) SetETag(t *T, newEtag string) {
+	setEtag(t, newEtag)
+}
+
+var _ accountSettingDefinition[struct{}] = accountSetting[struct{}]{}
+
+func makeSettingResource[T, U any](defn genericSettingDefinition[T, U]) *schema.Resource {
+	resourceSchema := common.StructToSchema(defn.SettingStruct(),
+		func(s map[string]*schema.Schema) map[string]*schema.Schema {
+			s["etag"].Computed = true
+			// Note: this may not always be computed, but it is for the default namespace setting. If other settings
+			// are added for which setting_name is not computed, we'll need to expose this somehow as part of the setting
+			// definition.
+			s["setting_name"].Computed = true
+			return s
+		})
+
+	createOrUpdate := func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient, setting T) error {
+		common.DataToStructPointer(d, resourceSchema, &setting)
+		var res string
+		switch defn := defn.(type) {
+		case workspaceSettingDefinition[T]:
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+			res, err = retryOnEtagError[T, string](func(setting T) (string, error) { return defn.Update(ctx, w, setting) }, setting, defn.SetETag)
+			if err != nil {
+				return err
+			}
+		case accountSettingDefinition[T]:
+			a, err := c.AccountClient()
+			if err != nil {
+				return err
+			}
+			res, err = retryOnEtagError(func(setting T) (string, error) { return defn.Update(ctx, a, setting) }, setting, defn.SetETag)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unexpected setting type: %T", defn)
+		}
+		d.SetId(res)
+		return nil
+	}
+
+	return common.Resource{
+		Schema: resourceSchema,
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			var setting T
+			return createOrUpdate(ctx, d, c, setting)
+		},
+		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			var res *T
+			switch defn := defn.(type) {
+			case workspaceSettingDefinition[T]:
+				w, err := c.WorkspaceClient()
+				if err != nil {
+					return err
+				}
+				res, err = retryOnEtagError(func(etag string) (*T, error) { return defn.Read(ctx, w, etag) }, d.Id(), func(req *string, newEtag string) { *req = newEtag })
```

Comment:
Can read fail on a EtagError?
From the docs:

The server will accept any etag and respond with a response which is at least as recent as the etag.

Reply:
Good point, removed need to retry in reads.